include/lbann/utils/dnn_lib/cudnn/convolution.hpp

```diff
@@ -0,0 +1,428 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+#ifndef LBANN_UTILS_DNN_LIB_CUDNN_CONVOLUTION_HPP_
+#define LBANN_UTILS_DNN_LIB_CUDNN_CONVOLUTION_HPP_
+
+#include "lbann/utils/cudnn.hpp"
+#include "lbann/utils/gpu/helpers.hpp"
+
+#include "utils.hpp"
+
+namespace lbann
+{
+
+/** @brief Which forward convolution algorithm to use. */
+enum class fwd_conv_alg
+{
+  IMPLICIT_GEMM,
+  IMPLICIT_PRECOMP_GEMM,
+  GEMM,
+  DIRECT,
+  FFT,
+  FFT_TILING,
+  WINOGRAD,
+  WINOGRAD_NONFUSED,
+};// enum class fwd_conv_alg
+
+/** @brief Which backward convolution algorithm to use. */
+enum class bwd_conv_alg
+{
+  GEMM, // "ALGO_0" in cuDNN (are the renamed correct?)
+  DIRECT, // "ALGO_1" in cuDNN
```

Comment:
If cuDNN exposes these algos as black boxes, I prefer not trying to divine what's inside:
```suggestion
  CUDNN_ALGO_0,
  CUDNN_ALGO_1,
```
If we do want more descriptive names, `IMPLICIT_GEMM_ATOMIC` and `IMPLICIT_GEMM` would be closer to the descriptions in the cuDNN docs.